packetbeat/protos/graphite/parser.go

```diff
@@ -0,0 +1,163 @@
+package graphite
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/streambuf"
+	"github.com/elastic/beats/packetbeat/protos/applayer"
+	"github.com/hydrogen18/stalecucumber"
+)
+
+type parser struct {
+	buf     streambuf.Buffer
+	config  *parserConfig
+	message *message
+
+	onMessage func(m *message) error
+}
+
+type parserConfig struct {
+	maxBytes int
+}
+
+type message struct {
+	applayer.Message
+
+	// indicator for parsed message being complete or requires more messages
+	// (if false) to be merged to generate full message.
+	isComplete bool
+	data       map[string]interface{}
+
+	// list element use by 'transactions' for correlation
+	next *message
+}
+
+// Error code if stream exceeds max allowed size on append.
+var (
+	ErrStreamTooLarge = errors.New("Stream data too large")
+)
+
+func (p *parser) init(
+	cfg *parserConfig,
+	onMessage func(*message) error,
+) {
+	*p = parser{
+		buf:       streambuf.Buffer{},
+		config:    cfg,
+		onMessage: onMessage,
+	}
+
+}
+
+func (p *parser) append(data []byte) error {
+	_, err := p.buf.Write(data)
+	if err != nil {
+		return err
+	}
+
+	if p.config.maxBytes > 0 && p.buf.Total() > p.config.maxBytes {
+		return ErrStreamTooLarge
+	}
+	return nil
+}
+
+func (p *parser) feed(ts time.Time, data []byte) error {
+	if err := p.append(data); err != nil {
+		return err
+	}
+
+	for p.buf.Total() > 0 {
+		if p.message == nil {
+			// allocate new message object to be used by parser with current timestamp
+			p.message = p.newMessage(ts)
+		}
+
+		msg, err := p.parse()
+		if err != nil {
+			return err
+		}
+		if msg == nil {
+			break // wait for more data
+		}
+
+		// reset buffer and message -> handle next message in buffer
+		p.buf.Reset()
+		p.message = nil
+
+		// call message handler callback
+		if err := p.onMessage(msg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *parser) newMessage(ts time.Time) *message {
+	return &message{
+		Message: applayer.Message{
+			Ts: ts,
+		},
+	}
+}
+
+func (p *parser) parse() (*message, error) {
+	// get the length of data in buffer
+	length := p.buf.Len()
+	// Read the entire buffer content
+	buf, err := p.buf.Collect(length)
+	if err == streambuf.ErrNoMoreBytes || length <= 2 {
+		return nil, nil
+	}
+
+	msg := p.message
+	msg.Size = uint64(p.buf.BufferConsumed())
+
+	isRequest := true
+	dir := applayer.NetOriginalDirection
+	pickledData := common.NetString(buf)
+	pickledDataIo := strings.NewReader(string(pickledData))
```

Comment:
why use strings.NewReader? Unpickl requires `io.Reader` only. You can use `bytes.Buffer` or `streambuf.StreamBuffer` here, without having to allocate and copy the complete payload.